[Hub Generated] Review request for Microsoft.Consumption to add version preview/2019-04-01-preview

ampravinr opened the pull request:
If you are a MSFT employee you can view your [work branch via this link](https://portal.azure-devex-tools.com/app/branch/ampravinr/azure-rest-api-specs/usage-consumption-Microsoft.Consumption-2019-04-01-preview...master).

### Contribution checklist:
- [ ] I have reviewed the [documentation](https://github.com/Azure/azure-rest-api-specs#basics) for the workflow.
- [ ] [Validation tools](https://github.com/Azure/azure-rest-api-specs/blob/master/documentation/swagger-checklist.md#validation-tools-for-swagger-checklist) were run on swagger spec(s) and have all been fixed in this PR.

ampravinr committed:
Adds base for updating Microsoft.Consumption from version stable/2019-01-01 to version 2019-04-01-preview

ampravinr committed:
Updates readme

ampravinr committed:
Updates API version in new specs and examples

ampravinr committed:
Usage Details API Contract changes for 2019-04-01-preview

Usage Details API Contract changes
- Usage Details model changes
- Usage details sample changes for the new model
- Usage Details download operation details
- Sample for usage details download
- Removed used oprerations and samples

ampravinr committed:
Usage Details API contract changes

- Added new samples for the query parameters
- Changed the "in" from path to query since this is optional query parameter.

ampravinr committed:
Usage Details API contract changes for 2019-04-01-preview

Included the samples in the consumption list

ampravinr committed:
Usage Details API contract changes for 2019-04-01-preview

- changed the values in samples.

ampravinr committed:
Usage API contract changes

- Change in samples

ampravinr committed:
Update UsageDetailsExpand.json

- changed the sample values

ampravinr committed:
Update UsageDetailsExpand.json

Changed the sample value

ampravinr committed:
Update consumption.json

Upated the sample file name in consumption.json

ampravinr committed:
Scope parameters added for the samples

- added the scope parameters for the samples

ampravinr committed:
Scope parameter fix

Scope parameter fixed in samples

ampravinr committed:
Scope parameters

Scope parameters added to the samples

ampravinr committed:
changed the scope parameter

added scope parameter to the samples

ampravinr committed:
Correcting scope parameter

Scope parameter changes

ampravinr committed:
Update Tags.json

Tags scope changes

ampravinr committed:
Fixed download and mg aggregated cost issues

Fixed date and download issues

ampravinr committed:
UsageDetails_Download operation id change

UsageDetails_Download operation id change

AutorestCI commented:
# Automation for azure-sdk-for-python

AutorestCI commented:
# Automation for azure-sdk-for-js

azuresdkci commented:
Can one of the admins verify this patch?

AutorestCI commented:
# Automation for azure-sdk-for-ruby

AutorestCI commented:
# Automation for azure-sdk-for-go

adxsdknet commented:
# Automation for azure-sdk-for-net 
A PR has been created for you: 
https://github.com/azure/azure-sdk-for-net/pull/5778 
.NET SDK Commits: 
https://github.com/adxsdknet/azure-sdk-for-net/pull/5778/commits/a79e3e9fdbb1c582160823241387203cd4f7f295

AutorestCI commented:
# Automation for azure-sdk-for-java

nschonni commented on specification/consumption/resource-manager/Microsoft.Consumption/preview/2019-04-01-preview/consumption.json:
```diff
@@ -0,0 +1,3159 @@
+        }
+      }
+    },
+    "/{scope}/providers/Microsoft.Consumption/usagedetails/download": {
```
```suggestion
    "/{scope}/providers/Microsoft.Consumption/usageDetails/download": {
```

ampravinr replied:
Thanks Nick for pointing this out. I have made the changes.

ampravinr committed:
Changed the casing for usage downloads operations

Changed the casing for usage downloads operations as per review

ampravinr committed:
Usage Details API contract changes

- Removed ActualAndAmortizedCost query options
- Removed samples for - ActualAndAmortizedCost
- Removed boolean isAmortizedCost and isActualCost from the model response.
- Removed the isActualCost and isAmortizedCost from samples

ampravinr committed:
Usage Detail API contract changes

Removed the unused boolean flgs from samples.

ampravinr committed:
Usage Details API contract changes

- Added parameter for Usage Download as well.
- Added sample for usage downloads with parmaeter.

veronicagg commented:
Have you evaluated warnings reported at https://portal.azure-devex-tools.com/app/branch/ampravinr/azure-rest-api-specs/usage-consumption-Microsoft.Consumption-2019-04-01-preview...master/linter ?

veronicagg reviewed (changes requested):
Please review https://dev.azure.com/azure-sdk/public/_build/results?buildId=17699&view=logs&jobId=8d2f9c49-cd83-5a9d-f3fe-2fd651afa742&taskId=43324f1f-6724-5d61-785f-f0ac9fd256a1&lineStart=82&lineEnd=83&colStart=1&colEnd=1
and https://dev.azure.com/azure-sdk/public/_build/results?buildId=17699&view=logs&jobId=f10f5372-40ed-54fc-caa4-b1cb5bd1252b&taskId=d3b2b841-a93b-5fa5-47c3-d8c4f6eae6e3&lineStart=14&lineEnd=15&colStart=1&colEnd=1

ampravinr committed:
Usage Details API contract changes

- Removed the bad sample file for Charges (as per the api version we will return only single object) - https://dev.azure.com/azure-sdk/public/_build/results?buildId=17699&view=logs&jobId=8d2f9c49-cd83-5a9d-f3fe-2fd651afa742&taskId=43324f1f-6724-5d61-785f-f0ac9fd256a1&lineStart=82&lineEnd=83&colStart=1&colEnd=1

- Removed the unused sample files as per - https://dev.azure.com/azure-sdk/public/_build/results?buildId=17699&view=logs&jobId=f10f5372-40ed-54fc-caa4-b1cb5bd1252b&taskId=d3b2b841-a93b-5fa5-47c3-d8c4f6eae6e3&lineStart=14&lineEnd=15&colStart=1&colEnd=1

ampravinr commented:
> Have you evaluated warnings reported at https://portal.azure-devex-tools.com/app/branch/ampravinr/azure-rest-api-specs/usage-consumption-Microsoft.Consumption-2019-04-01-preview...master/linter ?

Yes @veronicagg . This is all existing which we would like to keep as is.

ampravinr commented:
> Please review https://dev.azure.com/azure-sdk/public/_build/results?buildId=17699&view=logs&jobId=8d2f9c49-cd83-5a9d-f3fe-2fd651afa742&taskId=43324f1f-6724-5d61-785f-f0ac9fd256a1&lineStart=82&lineEnd=83&colStart=1&colEnd=1
> and https://dev.azure.com/azure-sdk/public/_build/results?buildId=17699&view=logs&jobId=f10f5372-40ed-54fc-caa4-b1cb5bd1252b&taskId=d3b2b841-a93b-5fa5-47c3-d8c4f6eae6e3&lineStart=14&lineEnd=15&colStart=1&colEnd=1

@veronicagg , i have reviewed and made the chages. Could you please verify from your side now.

prem9in commented:
@flanakin fyi.

RyanBensonMSFT commented on specification/consumption/resource-manager/Microsoft.Consumption/preview/2019-04-01-preview/consumption.json:
```diff
@@ -0,0 +1,3146 @@
+        }
+      }
+    },
+    "/{scope}/providers/Microsoft.Consumption/usageDetails/download": {
```
This URL references an instance of usageDetails named download. This cannot be used the way you're intending. I think what you're trying to do here is download the usageDetail collection but there is no pattern in ARM for /actions on collections, only /actions on instances.
 
We can either include the download url in the response to usageDetails or add POST /{scope}/providers/Microsoft.Consumption/downloadUsageDetails

flanakin replied:
Fwiw, I had the same thought when I first saw this. That said, this already exists today on older API versions. Instead of changing this only to change it again in a few months, let's please keep this one as-is and fix this when we add download support to Microsoft.CostManagement/query.

RyanBensonMSFT replied:
I don't know if following bad patterns is a great excuse. This will flag your RP as being non compliant with the RPC.

https://armwiki.azurewebsites.net/rp_onboarding/RPCstrongvalidation.html

Which API are you referring too?

ms-premp replied:
@RyanBensonMSFT I am not following this feedback . What exactly is getting non-compliant with RPC ?
I see your point that at present in ARM, looks like there is no pattern with actions on collections.
Does this mean, since a pattern does not exist, we should not introduce it all. May be there was never a need for it and now we have a scenario for such actions.
The standard pattern for OData for adding action on collections is https://docs.microsoft.com/en-us/aspnet/web-api/overview/odata-support-in-aspnet-web-api/odata-v3/odata-actions#binding-an-action-to-an-entity-set
This makes sense for our users and it's a lot simpler to understand APIs, when we stick to standards. 
Please let me know, if I am missing something.

RyanBensonMSFT commented on specification/consumption/resource-manager/Microsoft.Consumption/preview/2019-04-01-preview/consumption.json:
```diff
@@ -0,0 +1,3146 @@
+        }
+      }
+    },
+    "UsageDetailProperties": {
```
There are very few commonalities between this model and the one defined in 2019-01-01. Are you depricating the old model or just introducing a large change through api-version?

RyanBensonMSFT commented on specification/consumption/resource-manager/Microsoft.Consumption/preview/2019-04-01-preview/consumption.json:
```diff
@@ -0,0 +1,3146 @@
+        "externalDocs": {
+          "url": "https://docs.microsoft.com/en-us/rest/api/consumption/"
+        },
+        "x-ms-examples": {
```
Looks like the following was removed:
          "ChargesListForEnrollmentAccount": {
            "$ref": "./examples/ChargesListForEnrollmentAccountFilterByStartEndDate.json"
          },
          "ChargesListByDepartment": {
            "$ref": "./examples/ChargesListForDepartmentFilterByStartEndDate.json"
          }

ampravinr replied:
Yup the Charges API returns only single object (not array/list) for the API and tool was throwing error for the wrong samples. So I removed them and keeping only the right samples.

https://dev.azure.com/azure-sdk/public/_build/results?buildId=17699&view=logs&jobId=8d2f9c49-cd83-5a9d-f3fe-2fd651afa742&taskId=43324f1f-6724-5d61-785f-f0ac9fd256a1&lineStart=82&lineEnd=83&colStart=1&colEnd=1

RyanBensonMSFT commented on specification/consumption/resource-manager/Microsoft.Consumption/preview/2019-04-01-preview/consumption.json:
```diff
@@ -0,0 +1,3146 @@
+            "$ref": "#/parameters/apiVersionParameter"
+          },
+          {
+            "$ref": "#/parameters/metricParameter"
```
Is this the replacement for $apply?

ampravinr replied:
its not replacement. We are introducing a new Query Param from this version onwards.

flanakin replied:
This changes the underlying metric we use to get data. The user has to pick which metric they want and get that data set. There isn't an option to get multiple, so it's not a $filter. (We thought about that and tried to start there, but it forced us to add more data and support a scenario we don't actually want to support, increasing surface area and introducing perf/scale risks due to the size of the request.)'

ryansbenson reviewed (commented):
I'm having a hard time following the intent of this change as its vastly different than the previous GA'd api version.

ampravinr commented on specification/consumption/resource-manager/Microsoft.Consumption/preview/2019-04-01-preview/consumption.json:
```diff
@@ -0,0 +1,3146 @@
+        "externalDocs": {
+          "url": "https://docs.microsoft.com/en-us/rest/api/consumption/"
+        },
+        "x-ms-examples": {
```
Yes, this is intended. From this version onwards we are not supporting the Group by for this API. So removing the samples.

ampravinr commented on specification/consumption/resource-manager/Microsoft.Consumption/preview/2019-04-01-preview/consumption.json:
```diff
@@ -0,0 +1,3146 @@
+        }
+      }
+    },
+    "DownloadUrl": {
```
This will be a breaking change to us since Ibizza is already using this (internally) and they are consuming the downloadUrl. We can change it may be in a later version with backward compatible. But for this version we would like to keep it like downloadUrl. Also this is the same model we are keeping in our 2018-11-01-preview pricesheets downlaod as well.
- https://github.com/Azure/azure-rest-api-specs/blob/0972d9b7ba5b90b1c730ae744bcc655414d825bb/specification/consumption/resource-manager/Microsoft.Consumption/preview/2018-11-01-preview/consumption.json

flanakin replied:
Let's please keep it as-is. We're actively working on a new API to replace this and can align with ARM async download standards as part of that. That said, is there a defined standard? I've seen different patterns from different RPs. I'd hate to align to one nonstandard format only to be asked to align to another a few months later.

ampravinr commented on specification/consumption/resource-manager/Microsoft.Consumption/preview/2019-04-01-preview/consumption.json:
```diff
@@ -0,0 +1,3146 @@
+            "$ref": "#/parameters/scopeUsageDetailsParameter"
+          },
+          {
+            "name": "$expand",
```
Yes that's true. We are removing them from this versions and the previous versions still supports them.

ampravinr commented on specification/consumption/resource-manager/Microsoft.Consumption/preview/2019-04-01-preview/consumption.json:
```diff
@@ -0,0 +1,3146 @@
+                "description": "To get the progress of the operation, call GET operation on the URL in Azure-AsyncOperation header field.",
+                "type": "string"
+              },
+              "OData-EntityId": {
```
We are following the Async-operation approach for the downloads. As per the ARM stardards I believe this is the approach where we will mention the status url here in the AsyncOperation and the results url in the location property.

ampravinr replied:
I have removed the addiotional property.

veronicagg reviewed (changes requested):
Couple of inline questions.

flanakin commented on specification/consumption/resource-manager/Microsoft.Consumption/preview/2019-04-01-preview/consumption.json:
```diff
@@ -0,0 +1,3146 @@
+            "$ref": "./examples/UsageDetailsDownloadByMetricUsage.json"
+          }
+        },
+        "operationId": "UsageDetailsList_Download",
```
The intent is to make it very clear this is about downloading a file as opposed to a typical JSON GET request.

veronicagg commented:
@ampravinr will you be making any changes based on the comments? are you following up with ARM team offline on their questions? Thanks!

ampravinr commented:
Yes. Currently I am working to accommodate the Enum change you suggested on the query string  to make it Enum. Once I make changes and pushed the changes I will try to resume the Arm review to see review it and get sign off.
Also we where trying to see whether we can accommodate the change Arm suggested for listing different models based on kind and for now we cannot accommodate that so will go ahead with this for the Preview version.
Thanks,
Praveen Radhakrishnan
________________________________
From: Veronica Giaudrone <notifications@github.com>
Sent: Friday, April 19, 2019 8:33:12 PM
To: Azure/azure-rest-api-specs
Cc: Praveen Radhakrishnan; Mention
Subject: Re: [Azure/azure-rest-api-specs] [Hub Generated] Review request for Microsoft.Consumption to add version preview/2019-04-01-preview (#5622)


@ampravinr<https://nam06.safelinks.protection.outlook.com/?url=https%3A%2F%2Fgithub.com%2Fampravinr&data=02%7C01%7CPraveen.Radhakrishnan%40microsoft.com%7Cc12d87d61f1c463c973e08d6c5063e72%7C72f988bf86f141af91ab2d7cd011db47%7C1%7C0%7C636913027945897093&sdata=qbyUFegl7QJ2tqXoZdz%2BVoYwwl8g0lOt69bHEObga3U%3D&reserved=0> will you be making any changes based on the comments? are you following up with ARM team offline on their questions? Thanks!

-
You are receiving this because you were mentioned.
Reply to this email directly, view it on GitHub<https://nam06.safelinks.protection.outlook.com/?url=https%3A%2F%2Fgithub.com%2FAzure%2Fazure-rest-api-specs%2Fpull%2F5622%23issuecomment-485007146&data=02%7C01%7CPraveen.Radhakrishnan%40microsoft.com%7Cc12d87d61f1c463c973e08d6c5063e72%7C72f988bf86f141af91ab2d7cd011db47%7C1%7C0%7C636913027945907102&sdata=h8W82TN7S5EtLUTBOS%2B3%2FWv%2FLieW8zUHxKY4uXJe5aM%3D&reserved=0>, or mute the thread<https://nam06.safelinks.protection.outlook.com/?url=https%3A%2F%2Fgithub.com%2Fnotifications%2Funsubscribe-auth%2FAJXK4GRFYJIDI2EP6Q5XHOTPRIUARANCNFSM4HFNDMHA&data=02%7C01%7CPraveen.Radhakrishnan%40microsoft.com%7Cc12d87d61f1c463c973e08d6c5063e72%7C72f988bf86f141af91ab2d7cd011db47%7C1%7C0%7C636913027945907102&sdata=xcvCtD16cmq5iG%2FDQNsvEC4UyLo8rte28AjqZka78dA%3D&reserved=0>.

ampravinr committed:
metricParameter changes

defined allowed values for the metricParameter enum

ampravinr committed:
Changed the samples for ValidTill

Changed the download samples accordingly

ampravinr commented on specification/consumption/resource-manager/Microsoft.Consumption/preview/2019-04-01-preview/consumption.json:
```diff
@@ -0,0 +1,3146 @@
+                "readOnly": true
+              }
+            }
+          },
```
I have made the changes. Could you please review and let me know if any issues.

veronicagg commented:
@ampravinr ok, please work with ARM team on the ARM changes, so they sign off on the PR. 
Regarding CI I see python SDK for example failing to generate with this error: System.InvalidOperationException: Values specified by 'enum' mismatch those specified by 'x-ms-enum' (name: 'metrictype'): 'daily', 'monthly' vs 'usage', 'actualcost', 'amortizedcost'. Please take a [look](https://dev.azure.com/azure-sdk/public/_build/results?buildId=20062&view=logs&jobId=957f396e-b28a-53c4-b5bb-b8c6d959bd2d&taskId=142feabb-6d02-5968-caa1-f5b3c04c5cae&lineStart=134&lineEnd=134&colStart=41&colEnd=224).

ampravinr committed:
Fixed the enum values for metric types

Fixed the enum values for metric types

ampravinr committed:
changed the sample to match the case

changed the sample to match the enum case

ampravinr commented:
> @ampravinr ok, please work with ARM team on the ARM changes, so they sign off on the PR.
> Regarding CI I see python SDK for example failing to generate with this error: System.InvalidOperationException: Values specified by 'enum' mismatch those specified by 'x-ms-enum' (name: 'metrictype'): 'daily', 'monthly' vs 'usage', 'actualcost', 'amortizedcost'. Please take a [look](https://dev.azure.com/azure-sdk/public/_build/results?buildId=20062&view=logs&jobId=957f396e-b28a-53c4-b5bb-b8c6d959bd2d&taskId=142feabb-6d02-5968-caa1-f5b3c04c5cae&lineStart=134&lineEnd=134&colStart=41&colEnd=224).

@veronicagg - thanks for pointing me the error. I have fixed it. Please review and let me know if any issues.

veronicagg commented:
I'm good from my side, approval pending on ARM team's signoff. Thanks!

ampravinr committed:
As per review removed the OData-EntityId

Removed additional property for OData-EntityId.

ampravinr committed:
Changed the examples

Removed Odata-EntityId from samples

ampravinr committed:
OperationId change for Downloads

Based on the Doc Preview generated we want to change how the name appears. So changing the operation name for downloads.

ms-premp commented:
@ravbhatnagar we are waiting on ARM Feedback on this PR. Please have a look. Our public announcements for this preview version is at present waiting on this swagger changes.

ravbhatnagar commented:
Two concerns around this were discussed - 1. RBAC 2. Activity Log + Alert story. With this kind of modeling of action over collection. ARM will not parse the authorization action correctly as the instance segment is missing. hence what gets put in activity log will be wrong. And similarly RBAC will not be over the action which is what this API is. But since this is tenant level, RP handles its own RBAC. And since its tenant level, such logs dont get shown in activity log today. So these concerns are mitigated. ARM should look at this pattern and see if it makes sense to support it. 
The RP has agreed that this approval is for preview only. We will have another discussion for the GA story.
Signing off from ARM side.

ampravinr commented:
@veronicagg - we have the Arm sign off. Could you please merge these changes.

ampravinr commented:
Thanks Veronica for all the help on this.

Thanks,
Praveen Radhakrishnan

From: Veronica Giaudrone <notifications@github.com>
Sent: Thursday, May 2, 2019 11:14 AM
To: Azure/azure-rest-api-specs <azure-rest-api-specs@noreply.github.com>
Cc: Praveen Radhakrishnan <Praveen.Radhakrishnan@microsoft.com>; Mention <mention@noreply.github.com>
Subject: Re: [Azure/azure-rest-api-specs] [Hub Generated] Review request for Microsoft.Consumption to add version preview/2019-04-01-preview (#5622)


Merged #5622<https://nam06.safelinks.protection.outlook.com/?url=https%3A%2F%2Fgithub.com%2FAzure%2Fazure-rest-api-specs%2Fpull%2F5622&data=01%7C01%7CPraveen.Radhakrishnan%40microsoft.com%7C4ccf3990d1f24ed02dcb08d6cf29fac1%7C72f988bf86f141af91ab2d7cd011db47%7C1&sdata=jWMxEYHzLy918eeZ3JG758r%2FkyLT6x6Y9xkiA%2BA4oVQ%3D&reserved=0> into master.

—
You are receiving this because you were mentioned.
Reply to this email directly, view it on GitHub<https://nam06.safelinks.protection.outlook.com/?url=https%3A%2F%2Fgithub.com%2FAzure%2Fazure-rest-api-specs%2Fpull%2F5622%23event-2315306404&data=01%7C01%7CPraveen.Radhakrishnan%40microsoft.com%7C4ccf3990d1f24ed02dcb08d6cf29fac1%7C72f988bf86f141af91ab2d7cd011db47%7C1&sdata=wmjaezRGp0so%2FlZ0ypkuUjFF%2BuU2HyNYTPSTPdEetFM%3D&reserved=0>, or mute the thread<https://nam06.safelinks.protection.outlook.com/?url=https%3A%2F%2Fgithub.com%2Fnotifications%2Funsubscribe-auth%2FAJXK4GQYOHC4LSREW562Q53PTMVPHANCNFSM4HFNDMHA&data=01%7C01%7CPraveen.Radhakrishnan%40microsoft.com%7C4ccf3990d1f24ed02dcb08d6cf29fac1%7C72f988bf86f141af91ab2d7cd011db47%7C1&sdata=HjLudPdVPdXx7Ugi6R2JJk6MHq76RdJgVzetKCbRt7Q%3D&reserved=0>.
